testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java

```diff
@@ -3297,6 +3297,27 @@ static void checkOverlaps(OverlapChecker c) {
         DECIMAL_OVERFLOW, true);
   }
 
+  /** Test case for <a href="https://issues.apache.org/jira/browse/CALCITE-6706">[CALCITE-6706]
+   * Checked arithmetic does not take effect in subqueries</a>. */
+  @Test void testCastOverflow() {
+    final SqlOperatorFixture f = fixture().withConformance(SqlConformanceEnum.BIG_QUERY);
+    f.checkFails("SELECT -CAST(-32768 AS SMALLINT)",
+        ".*Value 32768 does not fit in a SMALLINT", true);
+    f.checkFails("SELECT CAST(32768 AS SMALLINT)",
+        "Value 32768 out of range", true);
+    f.checkFails("SELECT -CAST(32768 AS SMALLINT)",
+        "Value 32768 out of range", true);
+    final SqlOperatorFixture f0 = fixture();
+    // This query does not fail if checked arithmetic is not used
+    f0.checkScalar("SELECT -CAST(-32768 AS SMALLINT)",
+        "-32768", "SMALLINT");
+    // The last two queries should fail in any conformance level
```

Comment:
Because `32768 AS SMALLINT` is an illegal expression? Can we add a statement to explain that the out-of-bounds are caused by the parameter type, not the final result type?

Reply:
I have pushed another commit which enhances this comment explaining why this fails in all conformance levels.